commit 22d2ac3b0cfdb035b587a2425787bbeb65933b3a (refs/changes/63/2463/2)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-08-25T19:15:13+00:00 (2 years, 1 month ago)
    
    OS-6288 mmap of /dev/null is not zone-aware
    Reviewed by: Robert Mustacchi <rm@joyent.com>
    Approved by: Robert Mustacchi <rm@joyent.com>

diff --git a/usr/src/uts/common/fs/specfs/specsubr.c b/usr/src/uts/common/fs/specfs/specsubr.c
index a5937dec2b..014bad1216 100644
--- a/usr/src/uts/common/fs/specfs/specsubr.c
+++ b/usr/src/uts/common/fs/specfs/specsubr.c
@@ -24,6 +24,7 @@
  */
 /*
  * Copyright (c) 2012 by Delphix. All rights reserved.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*	Copyright (c) 1983, 1984, 1985, 1986, 1987, 1988, 1989 AT&T	*/
@@ -416,8 +417,13 @@ specfind(dev_t dev, vtype_t type)
 			nvp = STOV(st);
 			if (nvp->v_type == type && st->s_commonvp != nvp) {
 				VN_HOLD(nvp);
-				mutex_exit(&stable_lock);
-				return (nvp);
+				/* validate vnode is visible in the zone */
+				if (nvp->v_path != NULL &&
+				    ZONE_PATH_VISIBLE(nvp->v_path, curzone)) {
+					mutex_exit(&stable_lock);
+					return (nvp);
+				}
+				VN_RELE(nvp);
 			}
 		}
 		st = st->s_next;
